src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HMACCommon.cs

```diff
@@ -29,42 +30,60 @@ public int HashSizeInBits
         {
             get
             {
-                return _hashProvider.HashSizeInBytes * 8;
+                return _hMacProvider.HashSizeInBytes * 8;
             }
         }
 
         public void ChangeKey(byte[] key)
         {
-            HashProvider oldHashProvider = _hashProvider;
-            _hashProvider = null;
+            if (key.Length > _blockSize)
+            {
+                // Perform RFC 2104, section 2 key adjustment.
+                if (_lazyHashProvider == null)
+                    _lazyHashProvider = HashProviderDispenser.CreateHashProvider(_hashAlgorithmId);
+                _lazyHashProvider.AppendHashData(key, 0, key.Length);
+                key = _lazyHashProvider.FinalizeHashAndReset();
+            }
+
+            HashProvider oldHashProvider = _hMacProvider;
+            _hMacProvider = null;
```

Comment:
Nit: why set this to null?  We set it to something else in the statement after next.